src/search_decoder.py

```diff
@@ -56,3 +59,128 @@ def forward(self, input_seq, input_length, max_length):
         # Returen collections of word tokens and scores
         return all_tokens, all_scores
 
+class BeamSearchDecoder(nn.Module):
+    """
+    Inputs:
+        input_seq: an input sequence of shape (input_seq length, 1)
+        input_length: a scalar input length tensor
+        max_length: a max_length to bound the response sentence length.
+
+    Outputs:
+        all_tokens: collections of words tokens
+        all_scores: collections of words scores
+    """
+    
+    def __init__(self, encoder, decoder, speaker_id):
+        super(BeamSearchDecoder, self).__init__()
+        self.encoder = encoder
+        self.decoder = decoder
+        self.speaker = speaker_id
+
+    class BeamSearchNode(object):
+        def __init__(self, hiddenstate, previousNode, wordId, logProb, length):
+            '''
+            :param hiddenstate:
+            :param previousNode:
+            :param wordId:
+            :param logProb:
+            :param length:
+            '''
+            self.h = hiddenstate
+            self.prevNode = previousNode
+            self.wordid = wordId
+            self.logp = logProb
+            self.leng = length
+
+        def eval(self, alpha=1.0):
+            reward = 0
+            # Add here a function for shaping a reward
+            return self.logp / float(self.leng - 1 + 1e-6) + alpha * reward
+
+    def forward(self, input_seq, input_length, max_length):
+        beam_width = 10
+        topk = 1  # how many sentence do you want to generate
+        
+        # decoding goes sentence by sentence
+        # for idx in range(target_tensor.size(0)):
+        batch_size = 1 # TODO: don't know how to calculate
+        for _ in range(batch_size):
```

Comment:
Not sure if we need to fix the `batch_size` and how to do this. @ddddwy

Reply:
I think we don't need to fix it, since the `BeamSearchDecoder` is excuted on chat mode, and we only feed one sentence into the decoder at a time in chat mode.

Reply:
Good to know, then I think you can get this code by `git pull upstream pull/18/head` and test the beam search with your model, see if it works.

Reply:
Okay, thanks!